packages/@aws-cdk/aws-s3/lib/notifications-resource/lambda-source/src/index.py

```diff
@@ -0,0 +1,84 @@
+import boto3  # type: ignore
+import json
+import urllib.request
+from typing import List
+
+s3 = boto3.client("s3")
+
+
+def handler(event: dict, context):
+    response_status = "SUCCESS"
+    error_message = ""
+    try:
+        props = event["ResourceProperties"]
+        bucket = props["BucketName"]
+        s3.put_bucket_notification_configuration(
+            Bucket=bucket,
+            NotificationConfiguration=get_config(
+                current_config=s3.get_bucket_notification_configuration(Bucket=bucket),
+                new_config=props["NotificationConfiguration"],
+                old_config=event.get("OldResourceProperties", {}).get("NotificationConfiguration", {}),
+                request_type=event["RequestType"],
+            ),
+        )
+    except Exception as e:
+        print("Failed to put bucket notification configuration:", e)
+        response_status = "FAILED"
+        error_message = f"Error: {str(e)}. "
+    finally:
+        submit_response(event, context, response_status, error_message)
+
+
+def get_config(current_config: dict, new_config: dict, old_config: dict, request_type: str) -> dict:
+    config: dict = {}
+    types = (
+        ("TopicConfigurations", "TopicArn"),
+        ("QueueConfigurations", "QueueArn"),
+        ("LambdaFunctionConfigurations", "LambdaFunctionArn"),
+    )
+    for config_type, target in types:
+        configs = current_config.get(config_type, [])
+        in_configs = new_config.get(config_type, [])
+        if request_type == "Create":
+            configs.extend(in_configs)
+            config[config_type] = configs
+        elif request_type == "Update":
+            configs = find_difference(configs, old_config.get(config_type, []), target)
+            configs.extend(find_difference(in_configs, configs, target))
```

Comment:
Why call `find_difference` again here? Supposedly the line above detected the external configuration, so all thats left to do is extend it with the incoming config. 

Otherwise, you are effectively deduping notifications on update - which is bad since it causes delete events to delete external notifications. 

Here is the test scenario:

**Existing config:** ['notification1']
**Old config**: ['notification2']
**Incoming config:** ['notification1', 'notification2']
**Event Type**: Update

Right now this will succeed, but should fail because a user is trying to configure a notification that already exists.

Reply:
@iliapolo I am just removing the entries that match the old. And then adding in the new. I am not sure how that is a problem. Also i am not sure why that should fail.

Reply:
When the new includes a notification that already exists - it should fail. Right now, with this code - it will not. Thats the problem.

Run the following integ test:

0. Create a bucket with 1 notification.
1. Import that bucket in a different stack.
2. Add a new notification.
3. Deploy
4. Add a notification that matches the original existing notification.
5. Deploy - works (this should fail)
6. Destroy - this now deletes all notifications, but one should remain.

Reply:
@iliapolo ok, i thought i had a test case like this in the test suite. But maybe i am missing something. 

I will look at the suite, it is almost like we are effectly just ignoring what was created before. So do we even need read permissions?

Reply:
@iliapolo  what happens when you:

When you setup the following CDK stack:

1. Create a new S3 bucket with 1 notification from your CDK stack
2. And then later change that CDK stack notification to points to a new AWS Lambda function ARN
Deploy (we do not intend to keep the old one)
3. Add so the current notification that matches the original existing notification.
4. Deploy This should not fail we should allow for you to modify your notifications
5. Destroy - this now deletes all notifications and the S3 bucket

Reply:
@iliapolo I resolved this, see test `test_update_should_fail` to show that this is covered to.